### ISSUE TITLE
Basic implementation "client to server" ActivityPub protocol

### ISSUE THREAD

**Issue by yvolk** [andstatus/andstatus#499]
As defined at https://www.w3.org/TR/activitypub/ , ActivityPub provides two layers:
- A server to server federation protocol (so decentralized websites can share information)
- A client to server protocol (so users, including real-world users, bots, and other automated processes, can communicate with ActivityPub using their accounts on servers, from a phone or desktop or web application or whatever)

The https://test.activitypub.rocks/ site contains a test suite that checks basic feature that are required for a client app to implement #ActivityPub.

Below is a list of corresponding steps/features (for the start, copied from that site), which we are implementing in AndStatus, together with their implementation status:

Client: Basic submission
- [ ] MUST Client discovers the URL of a user's outbox from their profile
- [ ] MUST Client submits activity by sending an HTTP post request to the outbox URL with the Content-Type of application/ld+json; profile="https://www.w3.org/ns/activitystreams"
- [ ] MUST Client submission request body is either a single Activity or a single non-Activity Object
- [ ] Client submission request is authenticated with the credentials of the user to whom the outbox belongs (this could be using an OAuth 2.0 bearer token or something else)

Client: Required properties
- [ ] MUST Client provides the object property when submitting the following activity types to an outbox: Create, Update, Delete, Follow, Add, Remove, Like, Block, Undo.
- [ ] MUST Client provides the target property when submitting the following activity types to an outbox: Add, Remove.

Client: Add targets on reply
Reply to a post with multiple recipients.
- [ ] SHOULD Before submitting a new activity or object, Client infers appropriate target audience by recursively looking at certain properties (e.g. `inReplyTo`, See Section 7), and adds these targets to the new submission's audience. (The client suggests audience targeting based on participants in the referenced thread)
- [ ] SHOULD The client also limits depth of recursion used to gather targeting.

Client: Accept header on object retrieval
Trigger the client to retrieve some remote object.
- [ ] MUST When retrieving objects, Client specifies an Accept header with the application/ld+json; profile="https://www.w3.org/ns/activitystreams" media type

**Comment by ned14** [comment]
Thanks for the tip. Reading through https://github.com/pump-io/pump.io/blob/master/API.md, as far as I can tell these are the main significant differences between pump.io and W3C ActivityPub:

- Use of OAuth instead of HTTP signatures to authenticate. In fairness. the W3C spec leaves it open ended, but any of the W3C ActivityPub implementations I've curled so far use the latter.
- pump.io appears to hardcode the REST endpoints, and with slightly different naming. The W3C spec leaves ./well-known/webfinger complete flexibility to specify the endpoints to use.
- MIME types appear to be more specific in the W3C spec. Moreover, if you don't specify the specific W3C MIME types in the Accept: header, pleroma and Mastodon appear to try 301 redirecting you to elsewhere.

Otherwise, they do indeed look quite similar. The major blocker is the OAuth support really. I'm a bit loathe to implement it, because it's fundamentally broken, where HTTP signatures are not. But if it gets me up and running with andStatus quicker, I guess I can live with it.

**Comment by yvolk** [comment]
@ned14 Regarding restpoints. In pump.io, they are not hardcoded. Although they are usually the same, and this may lead to the hardcoded implementations in clients. 
See the sample pump.io response with endpoints explicitly set: https://github.com/andstatus/andstatus/blob/master/app/src/androidTest/res/raw/pumpio_actor_lists_person.json

Recently I fixed hardcoded endpoints for pump.io in AndStatus (exactly for ActivityPub migration), see https://github.com/andstatus/andstatus/blob/master/app/src/main/java/org/andstatus/app/net/social/ActorEndpoints.java

**Comment by yvolk** [comment]
@ned14 I think that now you can start testing/working at a server-side of "client to server" ActivityPub protocol. AndStatus client is just updated to allow creation of a Social Network of ActivityPub type.
See more details and a link to the prebuilt alpha version of AndStatus in the top post of this thread.

As I wrote there, I reused OAuth 2.0 implementation that works in AndStatus for Mastodon (with bearer token...). Everything else - from existing Pump.io implementation. Plus added "application/ld+json..." as required. 
If you need additional changes to the Client registration + User authentication/authorization part, feel free to change existing code. Btw
* Client Registration is implemented in subclasses of org.andstatus.app.net.http.HttpConnection#registerClient (called from org.andstatus.app.account.AccountSettingsActivity.OAuthRegisterClientTask ) - one implementation works in Mastodon, another - in Pump.io...
* User authentication/authorisation - in org.andstatus.app.account.AccountSettingsActivity.OAuthAcquireRequestTokenTask and org.andstatus.app.account.AccountSettingsActivity.OAuthAcquireAccessTokenTask 

In order to move forward, past authorisation step, I need an account on a server that successfully authenticates/authorises a User...

**Comment by yvolk** [comment]
A couple of years ago we did the same OAuth implementation exercise with @Gargron for Mastodon, please see this post https://github.com/andstatus/andstatus/issues/419#issuecomment-251169056 and below.
In short, an example/the simplest client application is needed (with one empty activity, e.g. from an Android template...), using any libraries, that can, using hard-coded username and password, connect to your server and request Home timeline for the User.
So we will know that:

* Client registration works
* Authentication works
* At least one timeline request works (no need to parse it)
I then integrate this into AndStatus, so the next steps will be easier.
?!

**Comment by ned14** [comment]
Bah, looks like I'll have to implement OAuth 2.0 so. It looks complex and hard to get right. But I'll give it a go. Thanks for the feedback, you saved me work!

**Comment by autogestion** [comment]
@ned14  You can try copypasting from this repo https://github.com/autogestion/pubgate
all that steps are already implemented there) except oauthAuthorizationEndpoint. But I think token endpoint is enough

**Comment by yvolk** [comment]
@autogestion @ned Please note that in both Pump.io and Mastodon client-to-server OAuth implementations the Client application starts with Dynamic client registration (implemented differently though...). Something like https://tools.ietf.org/html/rfc7591#section-1.3 (see Dynamic Registration as part of https://openid.net/connect/ also)
Only after successful Client Registration, User Identification and Authorisation starts...

**Comment by kaniini** [comment]
pleroma.site (as well as any other Pleroma install) has ActivityPub C2S available through OAuth.  You will need to use the Mastodon API to obtain client secrets for OAuth.

**Comment by yvolk** [comment]
@kaniini I registered at pleroma.site as andstatus@pleroma.site and then tried to adapt my current ActivityPub implementation to it:
1. Client registration works at 'https://pleroma.site/api/v1/apps'
2. Retrieval of request token works at 'oauth/token'
3. Authorisation via Web View works at 'https://pleroma.site/oauth/authorize...' and access token received.
4. Then I request "whoami" point and at this step I cannot get JSON with a User info, but getting either HTML document or even 406 server error.
I tried these URLs:
4.1. statusCode:OK (200); url:'https://pleroma.site/api/whoami'; response:'<!DOCTYPE html><html lang…'
4.2. statusCode:OK (200); url:'https://pleroma.site/api/v1/whoami'; response:'<!DOCTYPE html><html lang…'; 
4.3. statusCode:CLIENT_ERROR (406); url:'https://pleroma.site/api/v1/accounts/verify_credentials'; authenticated; response:'{"errors":{"detail":"Internal server error…';

What should I do, where can I get more info on correct API?

**Comment by kaniini** [comment]
the correct location to make AP C2S calls is at /users/:username/inbox

**Comment by yvolk** [comment]
> the correct location to make AP C2S calls is at /users/:username/inbox

@kaniini  Thank you, after several attempts I figured out the correct full URL:
https://pleroma.site/users/AndStatus/inbox
- and got JSON response. Will parse it and return to You to move further, if you don't mind :-)

However, this URL, looking like an actor's inbox, is conceptually incorrect for the "whoami" call as even AntivityPub spec says: "The **inbox** is discovered through the inbox property of an actor's profile". This is why the "whoami" call should be the same for every user, and not the "inbox" of a concrete user/actor.
 I think :-)

_this_ is "whoami" call where the Actor receives his/her profile information, including URLs of endpoints!

?!

**Comment by yvolk** [comment]
> > You will need to use the Mastodon API to obtain client secrets for OAuth.
> 
> Ugh wait. What's the point of having a C2S standard if OAuth registration isn't even part of it? How was it intended to work? 🤔

Yes, ActivityPub refers to OAuth but doesn't mention the "registration" word explicitly.
1. ActivityPub specification refers to OAuth 2.0 and OAuth 2.0 bearer tokens as **a** way to authenticate client to server interactions (see descriptions of corresponding endpoints in the "Actor objects" section https://www.w3.org/TR/activitypub/#actor-objects
2. In the (non-normative) "B.1 Authentication and Authorization" section the specification also refers to the Social Web Community Group Authentication and Authorization best practices report ( https://www.w3.org/wiki/SocialCG/ActivityPub/Authentication_Authorization ), which in a "Client to Server" section says:
> ActivityPub clients authenticate against a server using OAuth 2.0 bearer tokens. ...Client IDs may be acquired out of band for some servers. They may also be acquired using RFC7591.

The RFC7591 ( https://tools.ietf.org/html/rfc7591 ) is "OAuth 2.0 Dynamic Client Registration Protocol".

So OAuth **Client Registration** is at least a part of the specification's best practices :-)

We do have working OAuth 1.0 and Dynamic Client registration for a federated network, that's very close to ActivityPub from a Client app point of view ( pump.io https://github.com/pump-io/pump.io/blob/master/API.md ). Plus we do have OAuth 2.0 e.g. in Mastodon. So I think that we just need to go only one step further to implement OAuth 2.0 + Dynamic Client registration + "whoami" point in order to have a "bootstrap" for the "Client to Server interactions", which **are** actually part of the ActivityPub spec.

This is how it will work.
?!

**Comment by autogestion** [comment]
Have to say, that this is a bit strange to use OAuth in federated network. Because it's a standard which allows to authorize using some centralized authority, and fediverse do not have one

**Comment by kaniini** [comment]
@yvolk 

my current Pleroma tree has /api/ap/whoami for this purpose.  I will deploy it on pleroma.site in a few hours.  will this work for you?

**Comment by yvolk** [comment]
> @yvolk
> 
> my current Pleroma tree has /api/ap/whoami for this purpose. I will deploy it on pleroma.site in a few hours. will this work for you?

Yes, will test further and let you know. Thank you.

**Comment by kaniini** [comment]
I deployed that endpoint on pleroma.site.

**Comment by kaniini** [comment]
The AP C2S whoami endpoint which allows for mapping OAuth credentials to an actor endpoint is now in Pleroma upstream.

`oauthAuthorizationEndpoint` and `oauthTokenEndpoint` will be added to actors too.

**Comment by yvolk** [comment]
> The AP C2S whoami endpoint which allows for mapping OAuth credentials to an actor endpoint is now in Pleroma upstream.
> 
> `oauthAuthorizationEndpoint` and `oauthTokenEndpoint` will be added to actors too.
@kaniini Frankly I don't understand, why these two endpoints may be needed, if I get them _after_ OAuth authentication procedure.
What worries me in our growing solution, are URLs of the two quite voluntary/hardcoded endpoints:
1. Client registration at https://pleroma.site/api/v1/apps
2. whoami endpoint at https://pleroma.site/api/ap/whoami
Both endpoints are unknown to a client application, visiting this host.

I think that
1.  we need to use more "standard" URL for the Client Registration, e.g. 
https://pleroma.site/register (just like in examples, although non-normative, in https://tools.ietf.org/html/rfc7591
2. Whoami endpoint URL should be added as a separate property to a metadata, returned by a server during Client registration
?!

**Comment by lanodan** [comment]
[2019-02-06 22:23:57] Yuri Volkov:
> > The AP C2S whoami endpoint which allows for mapping OAuth credentials to an actor endpoint is now in Pleroma upstream.
> > 
> > `oauthAuthorizationEndpoint` and `oauthTokenEndpoint` will be added to actors too.
> @kaniini Frankly I don't understand, why these two endpoints may be needed, if I get them _after_ OAuth authentication procedure.

An actor is public data, it would allow to give https://pleroma.site/users/kaniini (actor IRI) or kaniini@pleroma.site (actor WebFinger) to register.
Which have the nice addition of giving username+instance in one shot.

> What worries me in our growing solution, are URLs of the two quite voluntary/hardcoded endpoints:
> 1. Client registration at https://pleroma.site/api/v1/apps
> 2. whoami endpoint at https://pleroma.site/api/ap/whoami
> Both endpoints are unknown to a client application, visiting this host.
> 
> I think that
> 1.  we need to use more "standard" URL for the Client Registration, e.g. 
> https://pleroma.site/register (just like in examples, although non-normative, in https://tools.ietf.org/html/rfc7591
> 2. Whoami endpoint URL should be added as a separate property to a metadata, returned by a server during Client registration
> ?!

Why would the mastodon one not become the standard one for client registeration? It’s basically the de-facto one.
But maybe there should be something in nodeinfo (but that’s not strictly ActivityPub) about which authentification is supported on the server.

**Comment by autogestion** [comment]
@yvolk 
> Frankly I don't understand, why these two endpoints may be needed, if I get them _after_ OAuth authentication procedure.

Why it is so important to perform client registration first? Client could start with fetching user profile using webfinger. It is de-facto standard for activitypub servers and is used to fetch user profiles, when user from one server searches for user from another server
https://mastodon.social/.well-known/webfinger?resource=acct:autogestion@mastodon.social
https://pleroma.site/.well-known/webfinger?resource=acct:kaniini@pleroma.site
https://pubgate.autogestion.org/.well-known/webfinger?resource=acct:dev@pubgate.autogestion.org

User profile MAY contain
> `oauthAuthorizationEndpoint` and `oauthTokenEndpoint` will be added to actors too.
and it could be extended with attribute which contains url for dynamic registration. And there will be no need to standardize endpoints names

Still, this discussion looks like trying to stretch owl over the globe. OAuth is designed for cases when there is certain centralized server (which defines endpoints due to OAuth spec), not for chaotic federation

**Comment by lanodan** [comment]
[2019-02-07 19:39:05] Mykhailo Kushchenko:
> User profile MAY contain
> > `oauthAuthorizationEndpoint` and `oauthTokenEndpoint` will be added to actors too.
> and it could be extended with attribute which contains url for dynamic registration. And there will be no need to standardize endpoints names
> 
> Still, this discussion looks like trying to stretch owl over the globe. OAuth is designed for cases when there is certain centralized server (which defines endpoints due to OAuth spec), not for chaotic federation

Please do not make a fictionnal argument: OAuth is there is many (all deployed ones?) fediverse implementations (StatusNet, GnuSocial, Friendica, Mastodon, Pleroma, …) and other authentification methods have been:
- (HTTP) Basic Auth: which I would regard as quite bad when long-lived because you have to keep the password in plaintext
- OpenID: which is even worse than OAuth at centralisation and it completely leaks access to a provider
- IndieAuth?: which is basically a wrapper and involves way too many parties to work (extra: OpenPGP auth is/was broken for myself, probably because of Elliptic Curves)

**Comment by yvolk** [comment]
> Why it is so important to perform client registration first? Client could start with fetching user profile using webfinger. It is de-facto standard for activitypub servers and is used to fetch user profiles, when user from one server searches for user from another server
> https://mastodon.social/.well-known/webfinger?resource=acct:autogestion@mastodon.social ...

@autogestion This is just another way to discover a user profile. But not an ActivityPub way. I don't see any links to such a way in ActivityPub docs...

**Comment by yvolk** [comment]
> An actor is public data, it would allow to give https://pleroma.site/users/kaniini (actor IRI) or kaniini@pleroma.site (actor WebFinger) to register. Which have the nice addition of giving username+instance in one shot.

@lanodan The problem is that even if I know my user's webFingerId "AndStatus@pleroma.site", I don't know URI of the user's (Actor's) profile at the "pleroma.site" host.

@kaniini I succeeded to create an account in AndStatus via ActivityPub interface, read the user's profile via "whoami" endpoint, discover (URLs of) all endpoints and request the "inbox" (Home) timeline. 
Now I need to adapt source code to parse the timeline.

**Comment by yvolk** [comment]
@kaniini BTW, it looks like the endpoint in the user's profile 
""sharedInbox": "https:\/\/pleroma.site\/inbox"" 
leads to nowhere (I received HTML page instead of JSON timeline...). Could you look at this?

**Comment by lanodan** [comment]
[2019-02-10 10:19:36] Yuri Volkov:
> > An actor is public data, it would allow to give https://pleroma.site/users/kaniini (actor IRI) or kaniini@pleroma.site (actor WebFinger) to register. Which have the nice addition of giving username+instance in one shot.
> 
> @lanodan The problem is that even if I know my user's webFingerId "AndStatus@pleroma.site", I don't know URI of the user's (Actor's) profile at the "pleroma.site" host.
You actually do:
``curl -s https://pleroma.site/.well-known/webfinger?resource=acct:kaniini@pleroma.site | jq '.links[] | select( .type == "application/activity+json" )'``
```JSON
{
  "type": "application/activity+json",
  "rel": "self",
  "href": "https://pleroma.site/users/kaniini"
}
```

**Comment by autogestion** [comment]
@yvolk 

> But not an ActivityPub way. I don't see any links to such a way in ActivityPub docs...
It has one mention https://www.w3.org/wiki/ActivityPub_and_WebFinger which leads to this thread w3c/activitypub#194 . And this is more like fediverse standard, it is also used by diaspora and other protocols

> it looks like the endpoint in the user's profile ""sharedInbox": "https://pleroma.site/inbox""
It have to accept only POST requests from other servers. The idea of this endpoint, that if server have to deliver update to several users on one server, it could post it once in shared inbox, instead of posting several times to inbox of every target

**Comment by yvolk** [comment]
@lanodan @autogestion Thank you for examples and links on usage of WebFinger protocol (defined in https://tools.ietf.org/html/rfc7033 ). Reading the thread https://github.com/w3c/activitypub/issues/194 I've got that WebFinger really could be used in a case we know WebFingerId of a User and need to discover his/her profile in ActivityPub, but not as a general way to discover users' profiles...
Unfortunately, WebFinger doesn't reduce a number of hardcoded endpoints in a case of ActivityPub account creation in a client application (e.g. in AndStatus): we still need URLs of "Client Registration" and "Whoami" endpoints.
So I've hardcoded the endpoints for now (thinking about introducing alternative URLs...) and now I'm moving forward in implementing most needed APIs.

**Comment by yvolk** [comment]
> > it looks like the endpoint in the user's profile ""sharedInbox": "https://pleroma.site/inbox""
> 
> It have to accept only POST requests from other servers. The idea of this endpoint, that if server have to deliver update to several users on one server, it could post it once in shared inbox, instead of posting several times to inbox of every target

@autogestion  Not only. As ActivityPub spec says, this endpoint also serves another purpose: "Some servers would also like to display a list of all messages posted publicly to the "known network"" and "sharedInbox endpoints SHOULD also be publicly readable OrderedCollection objects containing objects addressed to the Public special collection. Reading from the sharedInbox endpoint MUST NOT present objects which are not addressed to the Public endpoint."

So this endpoint may provide a publicly readable timeline also.
@kaniini

**Comment by puckipedia** [comment]
to clarify stuff on Kroeg (which has a full C2S and S2S impl of ActivityPub):

- The endpoints can be set up to contain the OAuth urls, and i am going to add `me` into the token response for OAuth (ala IndieAuth), since i will never implement `/api/whoami`.
- the `sharedInbox` endpoint is readable, and indeed returns a list of messages posted to that `sharedInbox`, so like a type of 'global timeline' as far as that concept exists in Kroeg.
- please test with Kroeg (i will provide an account if needed) since it does many things that are 100% legal inside ActivityPub but that many things just don't like.
- there's a `proxyUrl` endpoint that can be used to retrieve remote objects, which will ensure proper handling of things like `inReplyTo` a private object that hasn't been inlined

i tried adding a Kroeg account (using the latest play store version) and it just ended up not doing anything after requesting `/api/whoami` ("MyAccount.Builder: Couldn't set the Username to: Empty")

**Comment by yvolk** [comment]
@kaniini While implementing timeline parsing I found two bugs already in Pleroma, please see https://git.pleroma.social/pleroma/pleroma/issues

**Comment by yvolk** [comment]
@puckipedia "whoami" endpoint is really important to make sure that a User successfully "logged into" his/her own account AND to discover all the User's endpoints.
I looked at https://indieauth.spec.indieweb.org/#authentication - it assumes that a User will type (copy/paste) his profile URL in order to log in. Not very user friendly :-) 
Getting the URL from "me" property is good, I like it ( https://indieauth.spec.indieweb.org/#access-token-response )
Please provide me with testing account at your server, I will test it also.

BTW, I will create next AndStatus build as soon as I will succeed to post my first Note via ActivityPub C2S API.

**Comment by kaniini** [comment]
https://git.pleroma.social/pleroma/pleroma/merge_requests/821 adds the two AP endpoints for OAuth, as well as `litepub:oauthRegistrationEndpoint`.

I can add a `litepub:oauthVerifyEndpoint` which would be your whoami endpoint, but it still seems kinda silly to me.  I don't know of any situation where you would be given oauth credentials for an account that is different than that you started the flow with.

**Comment by yvolk** [comment]
> I can add a `litepub:oauthVerifyEndpoint` which would be your whoami endpoint, but it still seems kinda silly to me. I don't know of any situation where you would be given oauth credentials for an account that is different than that you started the flow with.

@kaniini  Adding "whoami" endpoint to an Actor's profile seems silly to me also :-) Please add a link to it as "me" property of the OAuth access token response just  like shown in https://indieauth.spec.indieweb.org/#access-token-response and as @puckipedia suggested.

Please note the the whoami endpoint serves very important purpose, which the users profile doesn't: successful access to it really proves that a client application is correctly authenticated by a server. Accessing this endpoint after initial account creation in a client application allows a User to check if connection to a server works now.

**Comment by kaniini** [comment]
Yeah I like the idea of just returning who the actor is in the OAuth response.  I'll do that in a separate MR.

**Comment by autogestion** [comment]
there is one more thread about profile url discovering w3c/activitypub#20 where they talk about "follow your nose" approach. The idea, as I get it, is that profile could be discovered by next pattern  -  https + domain + @ + username. Mastodon already support this schema ( https://mastodon.social/@dansup ). Looks like it's proper ActivityPub way, instead of tweaking OAuth

**Comment by yvolk** [comment]
> there is one more thread about profile url discovering [w3c/activitypub#20](https://github.com/w3c/activitypub/issues/20) where they talk about "follow your nose" approach. The idea, as I get it, is that profile could be discovered by next pattern - https + domain + @ + username. Mastodon already support this schema ( https://mastodon.social/@dansup ). Looks like it's proper ActivityPub way, instead of tweaking OAuth

@autogestion I also read that thread, and as I wrote earlier in https://github.com/andstatus/andstatus/issues/499#issuecomment-462412681 I've got that WebFinger really could be used in a case we know WebFingerId of a User and need to discover his/her profile in ActivityPub, but not as a general way to discover users' profiles.

ActivityPub operated with Actor IDs, which aren't necessarily WebfingerIDs. So usage of this way is limited anyway.

As a side note: During analysis of messages from different networks (different instances of GNU Social and Mastodon...) using AndStatus app I found out that "UserName @ Host" is not always exactly "Webfinger ID" is the sense of the unique identifier: UserName is a "case sensitive name" and may be e.g. "AndStatus" (i.e. "andstatus" means another user...). Whereas "Webfinger ID" have to be case insensitive: I have to convert it to the lowercase in order to successfully identify the same Actor (user) in different networks...

**Comment by yvolk** [comment]
@autogestion In short: assuming that
1. A server provides in the OAuth access token response the "me" property with "URL of my profile"
looks to me a far more natural ActivityPub implementation than:
2. A server interprets a request with "https + domain + @ + username" query as a request for the URL of a profile for an Actor with "preferredUsername" (in ActivityPub terminology) = provided "username" and with "id" that could be dereferenced to the server's host...

**Comment by lanodan** [comment]
https://hostname/@:username is awfully non-standard (even de-facto, it’s only Mastodon) and isn’t ActivityPub at all as on Mastodon the actor IRI is https://hostname/users/:username, the /@:username route is for the profile web page.

**Comment by yvolk** [comment]
@kaniini @autogestion I made the simplest implementation of #ActivityPub C2S protocol in #AndStatus, allowing me to log into pleroma.site, read home timeline a get other user's info (i.e. read the Actor's profile based on his "id").
But I've got an error response from pleroma.site server on an attempt to send a Note, so I created another bug report in pleroma issues tracker: https://git.pleroma.social/pleroma/pleroma/issues/650

Alpha build of AndStatus is attached to this comment: https://github.com/andstatus/andstatus/issues/499#issue-389014088 , source code is available in the repository.
![send--activitypub-pleroma-001](https://user-images.githubusercontent.com/207716/52915159-9e841f00-32e1-11e9-90de-22503034bf69.png)

**Comment by kaniini** [comment]
The error is that you're sending to inbox instead of outbox.  Not a pleroma bug just a mistake :)

**Comment by yvolk** [comment]
@kaniini It's not so simple, please see my response: https://git.pleroma.social/pleroma/pleroma/issues/650#note_17234

**Comment by yvolk** [comment]
@kaniini Thank you for investigation. In #AndStatus v.47.02-alpha (attached to the first post of this thread) I can successfully post a Note to pleroma.site via #ActivityPub C2S API and see the Note published in AndStatus.
The latest page of inbox and outbox timelines is loaded successfully, (I didn't implement paged loading yet...).

**Comment by yvolk** [comment]
@kaniini As we discussed earlier here it would be good to **add "whoami" ("me") endpoint** to an OAuth response in order not to hardcode this endpoint.
As I see, pleroma.site doesn't send this endpoint yet: I checked both Client registration response (preferred place for the endpoint that is the same for all accounts) 
and Access token response.
At least I can easily extract that endpoint from both places. 
Please implement!

**Comment by kaniini** [comment]
https://git.pleroma.social/pleroma/pleroma/merge_requests/942 adds that response.

**Comment by yvolk** [comment]
@kaniini Thank you, I implemented corresponding change in AndStatus, waiting for your change to appear live on the pleroma.site

**Comment by kaniini** [comment]
This change has gone live on pleroma.site.

**Comment by yvolk** [comment]
@kaniini Thank you, the "me" property works! AndStatus v.47.06 (released yesterday) uses this already.

As another step forward I raised the issue: https://git.pleroma.social/pleroma/pleroma/issues/741 "ActivityPub C2S: Request to the URL of the conversation's ID returns blank HTML page instead of JSON conversation".
Please implement.

**Comment by yvolk** [comment]
@kanini I raised another issue to pleroma: "ActivityPub C2S: No way to request newer page of the CollectionPage" that is really a blocker even for the "Basic" ActivityPub C2S implementation: https://git.pleroma.social/pleroma/pleroma/issues/751

Please implement.

**Comment by yvolk** [comment]
Some time ago posting to pleroma.site stopped working properly due to changes at Pleroma side. I fixed this in AndStatus v.49.00, download from #456 

The only critical thing to implement is posting attached images. @kaniini How to do this via "ActivityPub C2S"?

**Comment by yvolk** [comment]
I create the issue in Pleroma on uploading images: https://git.pleroma.social/pleroma/pleroma/issues/1171

**Comment by swift2plunder** [comment]
Copied from my Mastodon posts...

> 
> I believe the workflow with discovery should be:
> 
>  - The user inputs their ID
>  - The client retrieves the actor (by way of webfinger if it's in user@domain form)
>  - The client uses the oauthTokenEndpoint and oauthAuthorizationEndpoint from the endpoints property of the actor, if present, to obtain tokens and grants (fallback to Masto API endpoints)
> 
> The token endpoint should be retrieved from the actor for each account before registering the client, even if the client install has registered another account on the same domain, in order to support per client instance workflows. When the server operator is providing stateless authorization, per client instance support is necessary to allow users to approve scopes individually

In response to "Follow your nose":

> The actor document is the SDF. I'm just taking the pragmatic approach of adding the relevant elements from the ActivityPub spec to the current implementation. So content negotiation on the actor id is the preferred method for retrieving the actor document and the actor is where the scent begins for discovering other endpoints

The AP Actor document isn't a great service description, but it's the one we have

Finally, a more correct and coherent statement of what I was trying to say on [https://banana.dog/web/statuses/102810770163813523](https://banana.dog/web/statuses/102810770163813523) is that client_secret is *not* required in the registration response, but client_secret_expires *is* required if a client_secret is provided. At first I thought it was funny that the client didn't include the secret when making a grant request, but I see that the client_secret is only used when the client is changing its *own* information

*My* use case is porting games from play by email to play by AP, where players usually have a close relationship with someone developing a custom client. I'm also thinking of Epicyon, a more privacy focused AP implementation from another developer that would probably benefit if this workflow were settled

**Comment by yvolk** [comment]
@swift2plunder I'm glad to hear about another server side implementation. Please let me know when I can create an account at your server and start testing, improving and extending implementation in AndStatus.

**Comment by lanodan** [comment]
> I created the issue in Pleroma on uploading images: https://git.pleroma.social/pleroma/pleroma/issues/1171

@yvolk : I case you didn't get the notification on pleroma's gitlab, I have implemented it about a week ago: https://git.pleroma.social/pleroma/pleroma/merge_requests/1706

**Comment by yvolk** [comment]
@lanodan Thank you, the ball is on my side now :-)

Please don't forget to implement/fix https://git.pleroma.social/pleroma/pleroma/issues/751 also "No way to request newer page of the Collection" :-)

**Comment by yvolk** [comment]
@lanodan I've looked into the change https://git.pleroma.social/pleroma/pleroma/merge_requests/1706/diffs and checked uptated "whoami" response at https://pleroma.site/api/whoami
And I see that the change is not present at the https://pleroma.site :-( 
(there is no "uploadMedia" endpoint there)
Please deploy your change so I could actually test while developing client code
?!

**Comment by yvolk** [comment]
In v.51.02 posting and downloading multiple attachments is implemented. Tested at https://queer.hacktivis.me/AndStatus?tab=statuses thanks to @lanodan 

Now I can successfully post and download posted note with attachments (tried one and two so far...), but these posts aren't visible in the Web site's timeline yet...
![Screenshot_20191110-174848](https://user-images.githubusercontent.com/207716/68546720-8c578d00-03ea-11ea-833d-651e8071e8a8.png)

**Comment by yvolk** [comment]
See current state of this implementation in the first comment of this thread.
The only feature that I didn't implement yet: Add and Remove actions, applicable e.g. to server-side collections (e.g. "Bookmarks". Unfortunately, no known ActivityPub C2S server implemented this yet.

**Comment by yvolk** [comment]
Further discussion regarding proper OAuth implementation for ActivityPub C2S is here: https://github.com/andstatus/andstatus/issues/528 (April 2021)
